shell/platform/android/io/flutter/embedding/engine/loader/FlutterLoader.java

```diff
@@ -0,0 +1,342 @@
+    public static final String PUBLIC_AOT_SHARED_LIBRARY_NAME =
+        FlutterLoader.class.getName() + '.' + AOT_SHARED_LIBRARY_NAME;
+    public static final String PUBLIC_VM_SNAPSHOT_DATA_KEY =
+        FlutterLoader.class.getName() + '.' + VM_SNAPSHOT_DATA_KEY;
+    public static final String PUBLIC_ISOLATE_SNAPSHOT_DATA_KEY =
+        FlutterLoader.class.getName() + '.' + ISOLATE_SNAPSHOT_DATA_KEY;
+    public static final String PUBLIC_FLUTTER_ASSETS_DIR_KEY =
+        FlutterLoader.class.getName() + '.' + FLUTTER_ASSETS_DIR_KEY;
+
+    // Resource names used for components of the precompiled snapshot.
+    private static final String DEFAULT_AOT_SHARED_LIBRARY_NAME = "libapp.so";
+    private static final String DEFAULT_VM_SNAPSHOT_DATA = "vm_snapshot_data";
+    private static final String DEFAULT_ISOLATE_SNAPSHOT_DATA = "isolate_snapshot_data";
+    private static final String DEFAULT_LIBRARY = "libflutter.so";
+    private static final String DEFAULT_KERNEL_BLOB = "kernel_blob.bin";
+    private static final String DEFAULT_FLUTTER_ASSETS_DIR = "flutter_assets";
+
+    // Mutable because default values can be overridden via config properties
+    private String aotSharedLibraryName = DEFAULT_AOT_SHARED_LIBRARY_NAME;
+    private String vmSnapshotData = DEFAULT_VM_SNAPSHOT_DATA;
+    private String isolateSnapshotData = DEFAULT_ISOLATE_SNAPSHOT_DATA;
+    private String flutterAssetsDir = DEFAULT_FLUTTER_ASSETS_DIR;
+
+    private static FlutterLoader instance;
+
+    @NonNull
+    public static FlutterLoader getInstance() {
+        if (instance == null) {
+            instance = new FlutterLoader();
+        }
+        return instance;
+    }
+
+    private boolean initialized = false;
+    @Nullable
+    private ResourceExtractor resourceExtractor;
+    @Nullable
+    private Settings settings;
+
+    /**
+     * Starts initialization of the native system.
+     * @param applicationContext The Android application context.
+     */
+    public void startInitialization(@NonNull Context applicationContext) {
+        startInitialization(applicationContext, new Settings());
+    }
+
+    /**
+     * Starts initialization of the native system.
+     * <p>
+     * This loads the Flutter engine's native library to enable subsequent JNI calls. This also
+     * starts locating and unpacking Dart resources packaged in the app's APK.
+     * <p>
+     * Calling this method multiple times has no effect.
+     *
+     * @param applicationContext The Android application context.
+     * @param settings Configuration settings.
+     */
+    public void startInitialization(@NonNull Context applicationContext, @NonNull Settings settings) {
+        if (Looper.myLooper() != Looper.getMainLooper()) {
+          throw new IllegalStateException("startInitialization must be called on the main thread");
+        }
+        // Do not run startInitialization more than once.
+        if (this.settings != null) {
```

Comment:
optional: I would float this above the thrown exception, just to avoid crashing unless we absolutely have to.

Reply:
Done.